marathon/src/main/scala/io/buoyant/marathon/v2/Api.scala

```diff
@@ -101,21 +112,90 @@ object Api {
   }
 }
 
-private class AppIdApi(client: Api.Client, apiPrefix: String)
+private class AppIdApi(client: Api.Client, apiPrefix: String, auth: Option[Api.Auth])
   extends Api
   with Closable {
 
   import Api._
 
+  var authToken: Option[String] = None
+
   def close(deadline: Time) = client.close(deadline)
 
-  def getAppIds(): Future[Api.AppIds] = {
-    val req = http.Request(s"$apiPrefix/apps")
-    Trace.letClear(client(req)).flatMap(rspToApps(_))
-  }
+  private[this] def refreshToken(): Future[String] =
```

Comment:
I think we have to do some tricky synchronization here.  If multiple requests come in at once, they'll all see the token as invalid or missing and all kick off their own auth requests.

I think that there are basically 3 states that the token can be in: absent, present, or pending.  This could be represented as an `Activity[Option[String]]`.  

If the token is present, we can just use it.  If the token is absent, we set the state to pending and kick off an auth request (using synchronization to make sure things don't change out from under us).  If the token is pending, we can just map on the activity to wait for the value.

State is hard!